DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c

```diff
@@ -797,3 +860,394 @@ MergeFirmwareImages (
+  @retval EFI_NOT_FOUND           FMAP, CBFS, or 'key_manifest.bin' could not be found.
+  @retval EFI_OUT_OF_RESOURCES    Could not allocate memory for the hash output or temporary buffers.
+  @retval EFI_ABORTED             Cryptographic hash calculation failed.
+**/
+EFI_STATUS
+EFIAPI
+GetOemRootKeyHash (
+  IN CONST VOID *Image,
+  IN UINTN ImageLen,
+  IN OUT UINT8 **OemRootKeyHash
+  )
+{
+  EFI_STATUS Status;
+  struct cbfs_file *KmFile;
+  struct cbfs_image Cbfs;
+  CONST Fmap *ImageFmap;
+  BtgPubKey *OemKey;
+  UINTN OemKeyLen;
+  UINT8 *HashInput;
+  UINTN HashInputLen;
+
+  if (!GetFmap (Image, ImageLen, &ImageFmap)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a(): failed to parse firmware\n",
+      __FUNCTION__
+      ));
+    return EFI_NOT_FOUND;
+  }
+
+  if (!GetCbfs (Image, ImageFmap, &Cbfs)) {
+    DEBUG ((DEBUG_ERROR, "%a(): failed to load CBFS\n", __FUNCTION__));
+    return EFI_NOT_FOUND;
+  }
+
+  KmFile = cbfs_get_entry (&Cbfs, "key_manifest.bin");
+  if (KmFile == NULL)
+    return EFI_NOT_FOUND;
+
+  Status = GetOemRootKeyFromKm (
+    CBFS_SUBHEADER(KmFile),
+    KmFile->len,
+    &OemKey,
+    &OemKeyLen
+  );
+
+  if (EFI_ERROR (Status))
+    return Status;
+
+  HashInputLen = (OemKey->KeySize / 8) + sizeof(OemKey->Exponent);
+  HashInput = AllocatePool (HashInputLen);
+
+  CopyMem (HashInput, OemKey->Modulus, OemKey->KeySize / 8);
+  CopyMem (HashInput + (OemKey->KeySize / 8), &OemKey->Exponent, sizeof(OemKey->Exponent));
+
+  *OemRootKeyHash = AllocatePool (48);
+  if (*OemRootKeyHash == NULL) {
+    FreePool (HashInput);
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  if (!Sha384HashAll (HashInput, HashInputLen, *OemRootKeyHash)) {
+    DEBUG ((DEBUG_ERROR, "%a(): Failed to generate OEM key SHA384 hash\n", __FUNCTION__));
+    FreePool (HashInput);
+    FreePool (*OemRootKeyHash);
+    *OemRootKeyHash = NULL;
+    return EFI_ABORTED;
+  }
+
+  FreePool (HashInput);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Checks if the two coreboot images are Boot Guard enabled and are using the
+  same OEM root key for signing.
+
+  @param[in] Current     Current firmware image
+  @param[in] Updated     Updated firmware image
+  @param[in] ImageSize   Size of updated firmware image
+
+  @return TRUE     If the two images are compatible with each other
+  @return FALSE    If the two images are incompatiblE
+**/
+BOOLEAN
+EFIAPI
+AreImageBtgKeysCompatible (
+  IN CONST VOID  *Current,
+  IN CONST VOID  *Updated,
+  IN CONST UINTN ImageSize
+  )
+{
+  EFI_STATUS  Status;
+  UINT8       *CurrentOemKeyHash, *UpdatedOemKeyHash;
+  BOOLEAN     KeysCompatible;
+
+  // Platform is unfused and unlocked, so different BtG key doesn't matter.
+  if (!IsPlatformFused () && !IsDescriptorLocked())
+    return TRUE;
+
+  // Boot Guard is not currently deployed.
+  if (!IsBootGuardEnabled (Current, ImageSize))
+    return TRUE;
+
+  // Going from BtG -> No BtG is not possible if the platform is fused.
```

Comment:
The comment is for an older version of the code which checked for fused state but not descriptor's lock state.  Now we can get here when:
- platform is fused, descriptor is locked (keys must match)
- platform is fused, descriptor is unlocked (won't happen with current configuration)
- platform is unfused, descriptor is locked (do we need to check BootGuard keys in this case?  is this even possible/relevant?)